lib/motors/can_processor.hpp

```diff
@@ -3,15 +3,16 @@
 #include <cstdint>
 
 #include <core/types.hpp>
+#include <io/can.hpp>
 
 namespace hyped::motors {
 
 class CanProcessor {
```

Comment:
Based on what I assume is going to happen, you will require an interface `ICanProcessor` in `can.hpp`. The class here should implement that interface. (You will probably want to have a collection like `std::unordered_map<uint32_t,std::vector<std::shared_ptr<ICanProcessor>>>` in the main `Can` instance.

I am doubtful whether `CanProcessor` is a good name after these changes have been made. For one, I would call the interface `ICanSubscriber`, because each subscriber listens to a particular set of IDs. This particular instance will then be responsible for what? Handling messages from the motors? So maybe it's just a `Controller`? Food for thought.

Reply:
What is the purpose of the collection?
I'm also not 100% sure the canprocessor class is needed at all tbh, my original thinking was that when `can.cpp` receives a message then `CanProcessor` would take that and make it into what we want to use, but probably that would just happen in `controller` anyway

Reply:
So the original point of `CanProcessor` was the following:
- you may have many different CAN devices with many different IDs
- some of them you are interested in, some of them you aren't interested in (because CAN messages get sent to everybody)
- for those that you are interested in, there are many different clients (e.g. motors vs bms)
- in order to handle this elegantly you may allow clients to tell you what they're interested in
- so motors may implement something called a `CanProcessor` and give it to `Can` saying "hi there, we would like to receive messages to ID 0xabcde" 
- `Can` says "okay, cool, I'll write that down"
- when `Can` receives a message, it checks the ID and finds all the `CanProcessor`s that registered interest

Now, you need the data structure to do the last step. Previously we hade a vector of these and just went through them in linear time. That's not happening on my watch.

Reply:
That makes sense. And because motors vs BMS will want to do different things when they receive a message it's an interface so there are different motor can processor and bms can processor and whatever else.

So when a message is received, the ID is checked against the map then it calls `canprocessor.newMessage()` or something similar for every canprocessor in the vector?

Reply:
Exactly. In your current interface it would be `processMessage`, I believe.